Handle rendering of URL Query Parameters in renderURL

atreya2011 opened the pull request:
Resolves #9 

TODO:
- [x] Create a TypeScript function in template to build URL query parameters
- [x] Handle nested message types.
- [x] Handle primitive repeated fields.

lyonlai commented on generator/template.go:
```diff
@@ -58,7 +58,7 @@ export type {{.Name}} = {
 {{- else }}
   static {{.Name}}(req: {{tsType .Input}}, initReq?: fm.InitReq): Promise<{{tsType .Output}}> {
-    return fm.fetchReq<{{tsType .Input}}, {{tsType .Output}}>(` + "`{{renderURL .}}`" + `, {...initReq, {{buildInitReq .}}})
+    return fm.fetchReq<{{tsType .Input}}, {{tsType .Output}}>(` + "`{{renderURL .}}`" + `,  {{buildURLQueryParams .}} {...initReq, {{buildInitReq .}}})
```
`buildURLQueryParams` in compile time doesn't have any knowledge on how to handle runtime payload properly as discussed in #9  I would say

atreya2011 replied:
Yea, and so I have been thinking about how to implement this and the only way seems to be:
* Add a TypeScript function within the template which builds the URL Query parameters during runtime by analyzing the request payload and URL Path.

Let me know what you think about this idea 🙏🏼

lyonlai replied:
a runtime function should do the work.

atreya2011 replied:
Thank you for your feedback! I do have an idea about how to implement the runtime function and will update this PR soon 🙇🏼

lyonlai commented on registry/message.go:
```diff
@@ -78,5 +74,17 @@ func (r *Registry) analyseMessage(fileData *data.File, packageName, fileName str
+	// build field name list to store in typeInfo
+	for _, field := range data.Fields {
+		if !field.IsExternal && !field.IsRepeated && !field.IsOneOfField {
+			fieldList = append(fieldList, field.Name)
```
wouldn't the `FieldList` here a little bit limited? wondering how would you handle nested type from external.

atreya2011 replied:
Yes after giving a lot of thought, an implementation like this won't know the nature of the request payload during runtime. So I will probably be scraping this implementation...

atreya2011 commented:
@lyonlai What I finally ended up doing was:

1. Create a TypeScript function in template to build URL query parameters that handles nested message types & primitive repeated fields along with optional fields.
2. First I pass the deeply nested objected and the list of fields present in the URL path to `renderURLSearchParams` function.
3. In this function, the deeply nested objected is flattened using the `flattenObject` function and then the query string is built using the rules present in the http.proto while excluding the list of fields present in the URL path as parameters.

It returns an empty string if there are not URL parameters to be rendered.

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+    [] as string[][]
+  );
+
+  return urlSearchParams.length > 0
```
wouldn't mind to have just `new URLSearchParams(urlSearchParams).toString()` here

atreya2011 replied:
Thank you! I will fix this as per your suggestion 🙇🏼

lyonlai commented on generator/template.go:
```diff
@@ -232,15 +317,22 @@ func renderURL(r *registry.Registry) func(method data.Method) string {
+		fieldList = strings.TrimRight(fieldList, ", ") + "]"
+
+		if !method.ClientStreaming && method.HTTPMethod == "GET" {
+			url += fmt.Sprintf("${fm.renderURLSearchParams(req, %s)}", fieldList)
```
fieldList is not needed.

atreya2011 replied:
`fieldList` is constructed here and used in `renderURLSearchParams` to remove the list of fields from the request object that is present in the URL path as parameters. This is because, `renderURLSearchParams` takes the whole request object as its input, which contains the key that may be present in the URL path as a parameter. So we need to ensure that fields in the request object that are present as URL path parameters are removed.

lyonlai replied:
`fieldList` as the name is quite confusing. a name like `fieldsInPath` will make the code reading better

atreya2011 replied:
Thank you. I will change it as you suggested 🙇🏼 `fieldsInPath` makes more sense 👍🏼

lyonlai reviewed (changes requested):
I like the flattenObject idea, this makes runtime to pickup the fields being set. And not to grow the generated asset size too much.

Suggestions have been made. Looking forward to it.

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+
+      const newPath = Array.isArray(deeplyNestedObject)
+        ? path + "." + key
+        : [path, key].filter(Boolean).join(".");
```
The `filter(Boolean)` removes the first `.` from the keys in the flattened object. 

For example, a request object like this:
```js
const req = {
  a: 10,
  postReq: {
    b: 11
  },
  c: [23, 25],
  extMsg: {
    d: 12
  }
};
```

Without `filter(Boolean)`:
```js
{
  .a: 10
  .c.0: 23
  .c.1: 25
  .extMsg.d: 12
  .postReq.b: 11
}
```

With `filter(Boolean)`:
```js
{
  a: 10
  c.0: 23
  c.1: 25
  extMsg.d: 12
  postReq.b: 11
}
```

lyonlai replied:
This is because the path's not pass down correctly in the first level, then it makes this `filter(Boolean)` required for every nested level, which's not ideal. better just from root level pass down the correct path.

`newPath = path ? [path, key].join(".") : key

atreya2011 replied:
Ah, I see... I have made a very naive mistake. Thanks for pointing this out 🙏🏼

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+      const newPath = Array.isArray(deeplyNestedObject)
+        ? path + "." + key
+        : [path, key].filter(Boolean).join(".");
+      const isObject = [
```
Sure I will check it out, thanks for your feedback! 🙇🏼

atreya2011 replied:
@lyonlai I did some more research and found a simpler one line way of checking if something is a plain object.

```js
const isPlainObject = Object.prototype.toString.call(value) === "[object Object]" ? true : false;
```

The following are results of calling ` Object.prototype.toString.call(value)` on various types of values.
```
{"foo": bar, "baz": {1, 2}} => [object Object]
{"foo": 1} => [object Number]
{"foo": undefined} => [object Undefined]
{"foo: [1, 2]} => [object Array]
{"foo: null} => [object Null]
```
etc...

Let me know what you think about this 🙏🏼

lyonlai replied:
Next time could you please also show the source of where you get the idea from? also good for the reviewer to verify the idea. 

the message is confusing, calling `Object.prototype.toString` won't get the result you posted. need to call the function like `Object.prototype.toString.call(1)` then will get the result.

Though it's simple, however 

```
class A {}
var a = new A()
Object.prototype.toString.call(a)
```

will get the result: `"[object Object]"`.  which makes this simple implementation not sufficient. 

Please make a copy of the `isPlainObject.js` as stated above. it covers quite a few edge cases.

atreya2011 replied:
Thank you for your quick feedback. My apologies for not noticing the above edge case. 
As suggested I will use the transpiled version `isPlainObject.js` as the function itself imports other ramda functions within it 🙇🏼

lyonlai replied:
yup, it imports other `ramda` functions for sure. but there aren't complex logic to replicate though. let's see

atreya2011 replied:
This is the final form of the `isPlainObj.js` (latest version) from `ramda-adjunct` because `ramda` library does not contain a function called `isPlainObject.js`. It imports a lot of helpers from `ramda` and those helpers import other helper functions as well. Is it possible to include the `ramda.min.js` as a separate generated file?

Link for source: https://github.com/char0n/ramda-adjunct/blob/master/src/isPlainObj.js


```js
import {
  pipe,
  type,
  identical,
  both,
  equals,
  toString,
  pathSatisfies,
  curryN,
  anyPass,
  complement,
  is,
  either,
  stubFalse
} from "ramda";

const isOfTypeObject = val => typeof val === "object";

/**
 * Checks if input value is `null`.
 *
 * @func isNull
 * @memberOf RA
 * @since {@link https://char0n.github.io/ramda-adjunct/0.1.0|v0.1.0}
 * @category Type
 * @sig * -> Boolean
 * @param {*} val The value to test
 * @return {boolean}
 * @see {@link RA.isNotNull|isNotNull}
 * @example
 *
 * RA.isNull(1); //=> false
 * RA.isNull(undefined); //=> false
 * RA.isNull(null); //=> true
 */
const isNull = equals(null);

/**
 * Checks if input value is complement of `null`.
 *
 * @func isNotNull
 * @memberOf RA
 * @since {@link https://char0n.github.io/ramda-adjunct/0.1.0|v0.1.0}
 * @category Type
 * @sig * -> Boolean
 * @param {*} val The value to test
 * @return {boolean}
 * @see {@link RA.isNull|isNull}
 * @example
 *
 * RA.isNotNull(1); //=> true
 * RA.isNotNull(undefined); //=> true
 * RA.isNotNull(null); //=> false
 */
const isNotNull = complement(isNull);

/* eslint-disable max-len */
/**
 * Checks if value is object-like. A value is object-like if it's not null and has a typeof result of "object".
 *
 * @func isObjLike
 * @aliases isObjectLike
 * @memberOf RA
 * @since {@link https://char0n.github.io/ramda-adjunct/0.5.0|v0.5.0}
 * @category Type
 * @sig * -> Boolean
 * @param {*} val The value to test
 * @return {boolean}
 * @see {@link RA.isNotObjLike|isNotObjLike}, {@link RA.isObj|isObj}, {@link RA.isPlainObj|isPlainObj}
 * @example
 *
 * RA.isObjLike({}); //=> true
 * RA.isObjLike([]); //=> true
 * RA.isObjLike(() => {}); //=> false
 * RA.isObjLike(null); //=> false
 * RA.isObjLike(undefined); //=> false
 */
/* eslint-enable max-len */
const isObjLike = curryN(1, both(isNotNull, isOfTypeObject));

let GeneratorFunction = null;
let legacyCheck = null;
try {
  GeneratorFunction = new Function("return function* () {}")().constructor; // eslint-disable-line no-new-func
  legacyCheck = is(GeneratorFunction);
} catch (e) {
  legacyCheck = stubFalse;
}

/**
 * Checks if input value is `Generator Function`.
 *
 * @func isGeneratorFunction
 * @memberOf RA
 * @since {@link https://char0n.github.io/ramda-adjunct/0.5.0|v0.5.0}
 * @category Type
 * @sig * -> Boolean
 * @param {*} val The value to test
 * @return {boolean}
 * @see {@link RA.isFunction|isFunction}, {@link RA.isAsyncFunction|isAsyncFunction}, {@link RA.isNotGeneratorFunction|isNotGeneratorFunction}
 * @example
 *
 * RA.isGeneratorFunction(function* test() { }); //=> true
 * RA.isGeneratorFunction(null); //=> false
 * RA.isGeneratorFunction(function test() { }); //=> false
 * RA.isGeneratorFunction(() => {}); //=> false
 */
const isGeneratorFunction = curryN(
  1,
  either(
    pipe(
      type,
      identical("GeneratorFunction")
    ),
    legacyCheck
  )
);

/**
 * Checks if input value is `Async Function`.
 *
 * @func isAsyncFunction
 * @memberOf RA
 * @since {@link https://char0n.github.io/ramda-adjunct/0.5.0|v0.5.0}
 * @category Type
 * @sig * -> Boolean
 * @param {*} val The value to test
 * @return {boolean}
 * @see {@link RA.isFunction|isFunction}, {@link RA.isNotAsyncFunction|isNotAsyncFunction}, {@link RA.isGeneratorFunction|isGeneratorFunction}
 * @example
 *
 * RA.isAsyncFunction(async function test() { }); //=> true
 * RA.isAsyncFunction(null); //=> false
 * RA.isAsyncFunction(function test() { }); //=> false
 * RA.isAsyncFunction(() => {}); //=> false
 */
const isAsyncFunction = curryN(
  1,
  pipe(
    type,
    identical("AsyncFunction")
  )
);

/**
 * Checks if input value is `Function`.
 *
 * @func isFunction
 * @memberOf RA
 * @since {@link https://char0n.github.io/ramda-adjunct/0.5.0|v0.5.0}
 * @category Type
 * @sig * -> Boolean
 * @param {*} val The value to test
 * @return {boolean}
 * @see {@link RA.isNotFunction|isNotFunction}, {@link RA.isAsyncFunction|isNotAsyncFunction}, {@link RA.isGeneratorFunction|isGeneratorFunction}
 * @example
 *
 * RA.isFunction(function test() { }); //=> true
 * RA.isFunction(function* test() { }); //=> true
 * RA.isFunction(async function test() { }); //=> true
 * RA.isFunction(() => {}); //=> true
 * RA.isFunction(null); //=> false
 * RA.isFunction('abc'); //=> false
 */
const isFunction = anyPass([
  pipe(
    type,
    identical("Function")
  ),
  isGeneratorFunction,
  isAsyncFunction
]);

const isObject = pipe(
  type,
  identical("Object")
);
const isObjectConstructor = pipe(
  toString,
  equals(toString(Object))
);
const hasObjectConstructor = pathSatisfies(
  both(isFunction, isObjectConstructor),
  ["constructor"]
);

/* eslint-disable max-len */
/**
 * Check to see if an object is a plain object (created using `{}`, `new Object()` or `Object.create(null)`).
 *
 * @func isPlainObj
 * @aliases isPlainObject
 * @memberOf RA
 * @since {@link https://char0n.github.io/ramda-adjunct/0.5.0|v0.5.0}
 * @category Type
 * @sig * -> Boolean
 * @param {*} val The value to test
 * @return {boolean}
 * @see {@link RA.isNotPlainObj|isNotPlainObj}, {@link RA.isObjLike|isObjLike}, {@link RA.isObj|isObj}
 * @example
 *
 * class Bar {
 *   constructor() {
 *     this.prop = 'value';
 *   }
 * }
 *
 * RA.isPlainObj(new Bar()); //=> false
 * RA.isPlainObj({ prop: 'value' }); //=> true
 * RA.isPlainObj(['a', 'b', 'c']); //=> false
 * RA.isPlainObj(Object.create(null); //=> true
 * RA.isPlainObj(new Object()); //=> true
 */
/* eslint-enable max-len */
const isPlainObj = curryN(1, val => {
  if (!isObjLike(val) || !isObject(val)) {
    return false;
  }

  const proto = Object.getPrototypeOf(val);

  if (isNull(proto)) {
    return true;
  }

  return hasObjectConstructor(proto);
});

export default isPlainObj;
```

lyonlai replied:
it seems complex at the first glance. but essentially the logic is quite simple to copy  and the imported ramda functions are not necessary. 

isObject
```
a => Object.prototype.toString.call({a: 1}).slice(8, -1) === 'Object'
```

isObjLike
```
a => a !== null && isObject
```

hasObjectConstructor
```
// give proto.constructor as input
typeof input === 'function' && input === Object.prototype.constructor
```

atreya2011 replied:
> it seems complex at the first glance. but essentially the logic is quite simple to copy and the imported ramda functions are not necessary.
> 
> isObject
> 
> ```
> a => Object.prototype.toString.call({a: 1}).slice(8, -1) === 'Object'
> ```
> 
> isObjLike
> 
> ```
> a => a !== null && isObject
> ```
> 
> hasObjectConstructor
> 
> ```
> // give proto.constructor as input
> typeof input === 'function' && input === Object.prototype.constructor
> ```


I see. I will create a seperate `isObject` function based on your feedback code 🙇🏼

atreya2011 replied:
@lyonlai 
Based on your feedback, I have created a `isPlainObject` function.
```ts
function isPlainObject(val: unknown): boolean {
  const isObject =
    Object.prototype.toString.call(val).slice(8, -1) === "Object";
  const isObjLike = val !== null && isObject;

  if (!isObjLike || !isObject) {
    return false;
  }
  
  const proto = Object.getPrototypeOf(val);

  const hasObjectConstructor =
    typeof proto === "object" &&
    proto.constructor === Object.prototype.constructor;

  console.log(typeof proto === "function");

  return hasObjectConstructor;
}
```
I tried this `typeof input === 'function' && input === Object.prototype.constructor`. However, this returns `false` even if given input is an object.
Hence I modified it to below logic because according to my understanding Object.prototype should not be a function
```js
  const hasObjectConstructor =
    typeof proto === "object" &&
    proto.constructor === Object.prototype.constructor;
```

 Let me know if I have missed anything 🙇🏼

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+    (acc: T, key: string): T => {
+      const value = deeplyNestedObject[key];
+
+      const newPath = Array.isArray(deeplyNestedObject)
```
Yes, I did consider this initially but the implementation was getting complicated... let me try it again 🙇🏼

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+function flattenObject<T extends Record<string, unknown>>(
+  deeplyNestedObject: T,
+  path: string = ""
+): T {
```
Thank you for your feedback. Yes, this makes sense, maybe `type GenericObject = Record<string, unknown>` ? 
Let me know your opinion on this 🙇🏼

lyonlai replied:
it's not quite a `GenericObject` though. let's call it `RequestPayload`.

atreya2011 replied:
Sure. That sounds good. Thank you 🙏🏼

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+ */
+export function renderURLSearchParams<T extends Record<string, unknown>>(
+  deeplyNestedObject: T,
+  urlPathParams: string[] = []
```
urlPathParams is required here because `renderURLSearchParams` takes the whole request object as its input which also contains the key that may be present in the URL path as a parameter. So we need to ensure that fields in the request object that are present as URL path parameters are removed.

For example. Lets take the following RPC with endpoint:
```proto
message HTTPGetWithURLSearchParamsRequest {
  int32 a = 1;
  PostRequest post_req = 2;
  repeated int32 c = 3;
  ExternalMessage ext_msg = 4;
}

message HTTPGetWithURLSearchParamsResponse {
  int32 url_search_params_result = 1;
}

service CounterService {
  rpc HTTPGetWithURLSearchParams(HTTPGetWithURLSearchParamsRequest) returns (HTTPGetWithURLSearchParamsResponse) {
    option (google.api.http) = {
      get: "/api/query/{a}"
    };
  }
}
```

`renderURLSearchParams` takes the following whole request object as the input which contains the field `a` that is also present as the URL path parameter. 

```js
const req = {
  a: 10,
  postReq: {
    b: 11
  },
  c: [23, 25],
  extMsg: {
    d: 12
  }
};
```

Therefore we need to ensure that we remove field `a` from the string that is generated as a response of `renderURLSearchParams`

lyonlai replied:
cool. understand now. the confusion comes from `fieldList` as per comment later down the change

atreya2011 replied:
Thank you for understanding 🙇🏼

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+
+      // if key does not contain only one numeric index as the last element
+      // then it is an array of objects
+      if (parts.length === index + 1 || index === -1) {
```
Thank you 🙏🏼

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+        // values should only be primitive types and key should not be
+        // present in the url path as a parameter
+        const value = flattenedObject[key];
+        if (
```
Perhaps we can restrict `unknown` to `Array<Primitive> | Primitive` where `type Primitive = string | boolean | number`. Let me know what you think about this 🙇🏼

lyonlai replied:
sounds good. let's give it a try.

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+): string {
+  const flattenedObject = flattenObject(deeplyNestedObject);
+
+  const urlSearchParams = Object.keys(flattenedObject).reduce(
```
I have used `reduce` here because renderURLSearchParams takes the whole request object as its input, which contains the key that may be present in the URL path as a parameter. So we need to ensure that fields in the request object that are present as URL path parameters are removed.

lyonlai replied:
once the complex logic is removed. I guess you could just chain the calls like 

`Object.keys(flattenedObject).filter(keysInPath).map()`

atreya2011 replied:
Aaah! Indeed. Thank you for your suggestion. I will try this out 👍🏼

atreya2011 commented:
@lyonlai 
As per your suggestion, I tried this `Object.keys(flattenedObject).filter(keysInPath).map()`  but unfortunately, it becomes complicated to return an array of type `string[][]` to be used in `new URLSearchParams()`, so I simplified the `Object.reduce` function.

For example, lets take the following request payload where we filter out key `a` assuming it is present in the path as a parameter:
```js
const req = {
  a: 10,
  postReq: {
    b: 11
  },
  c: [23, 25],
  extMsg: {
    d: 12
  }
};
```

After flattening and filtering key `a` it becomes like this:

```js
{
  "postReq.b": 11,
  c: [23, 25],
  "extMsg.d": 12
};
```

The above flattened request payload needs to be converted like this to use as input in `new URLSearchParams()`:
```js
[["postReq.b", 11], ["c", 23], ["c", 25], ["extMsg.d", 12]]
```

However, it becomes complicated to use just Array.map on the filtered and flattened to return an array of type `string[][]` like above.
So I have simplified the reduce function instead. 

Looking forward to your feedback 🙇🏼

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +195,97 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+        value.every(v => isPrimitive(v)) &&
+        value.length > 0;
+
+      switch (true) {
```
`switch(true)` seems quite odd to me. you could turn this block into something like the following

```
let objectToMerge = {}

if (isObject) {
  objectToMerge = flattenRequestPayload(value as RequestPayload, newPath)
} else if (isPrimitive(value) || isNonEmptyPrimitiveArray) {
  objectToMerge = {[newPath]: value}
}

return {...acc, ...objectToMerge}
```

atreya2011 replied:
Thank you for your feedback. I will revise this as per your suggestion 🙇🏼

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +195,97 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+      return acc;
+    },
+    {} as T
+  ) as Record<string, Primitive | Array<Primitive>>;
```
Could we name `Record<string, Primitive | Array<Primitive>>` to `FlattenedRequestPayload`?

atreya2011 replied:
Thank you. Will do so 👍🏼

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +195,97 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+      const value = requestPayload[key];
+      const newPath = path ? [path, key].join(".") : key;
+
+      const isObject =
```
`isObject` as per commented might not cover the edge cases.

atreya2011 replied:
Thank you for your feedback and I understand your concern. I have pasted the final form of `isPlainObj.js` from `ramda adjunct` library. Let me know what you think of this. My comment is here
https://github.com/grpc-ecosystem/protoc-gen-grpc-gateway-ts/pull/11#discussion_r620745005

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +195,97 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+ * @param  {unknown} value
+ * @return {boolean}
+ */
+function isPrimitive(value: unknown): boolean {
```
`["string", "number", "boolean"].some(t => typeof value === t)`

atreya2011 replied:
Thank you for this simpler implementation suggestion 🙇🏼

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +195,97 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+  path: string = ""
+): Record<string, Primitive | Array<Primitive>> {
+  return Object.keys(requestPayload).reduce(
+    (acc: T, key: string): T => {
```
there's also case where the key could be assigned to null and it will present in the keys because it's set.

`Object.keys({a: null}).forEach(k => console.log("key", k))` => `key a`

`Object.keys({a: undefined}).forEach(k => console.log("key", k))` => `key a`

You might want to filter out the null/undefined case here.

lyonlai replied:
^^ also could be added to the test case you have

lyonlai replied:
in fact, wouldn't the falsy cases the default value anyway? can you verify?  that might also save the length of the url but it will lose some readability on the URL. <- as in, `"I've set the value to 0/false, but why it doesn't show up in the url?`. It might just worth a sentense in the readme for this.

atreya2011 replied:
The function `isPlainObj.js` will ensure that undefined and null are filtered out. I will add a test case to check this as well 🙇🏼

atreya2011 replied:
I will add tests to verify falsy cases such as a `0/false`. 👍🏼

lyonlai replied:
oh, right. I see. ah. all good. yeah. 0 & false case then

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +195,97 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+    (acc: string[][], key: string): string[][] => {
+      // key should not be present in the url path as a parameter
+      const value = flattenedRequestPayload[key];
+      if (!urlPathParams.find(f => f === key)) {
```
can we flatten the nested if like the following?

```
if (!urlPathParams.find(f => f === key)) {
  return acc
}

return Array.isArray(value)
? ...
: ...

```

atreya2011 replied:
Ah yes, I was just thinking to refactor it that way. Thank you for your suggestion 🙇🏼

lyonlai replied:
can we revert the logic here? it's just easier to read without the negative conditions in if.

```
if (urlPathParams.find(f => f === key)) {
  return acc
}
```

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +195,97 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+    [] as string[][]
+  );
+
+  return "?" + new URLSearchParams(urlSearchParams).toString();
```
there could be another case where in the url there's already a hardcoded part of query string specified, e.g. `test/{a}?predefined-query=a`. in that case  this is an `&` rather than a `?`.  I would suggest not worry about the `?` here and the leave it to the generator logic.

atreya2011 replied:
In the final generated file, the renderURLSearchParams is part of the Path string and therefore I don't think there will be a case where there's already a hardcoded query string specified.
`/api/query/${req["a"]}${fm.renderURLSearchParams(req, ["a"])}`

`new URLSearchParams()` does not generate a `?` and the test fails if we do not prepend a `?` to the final generated URL Search Params.

```proto
  static HTTPGetWithURLSearchParams(req: HTTPGetWithURLSearchParamsRequest, initReq?: fm.InitReq): Promise<HTTPGetWithURLSearchParamsResponse> {
    return fm.fetchReq<HTTPGetWithURLSearchParamsRequest, HTTPGetWithURLSearchParamsResponse>(`/api/query/${req["a"]}${fm.renderURLSearchParams(req, ["a"])}`, {...initReq, method: "GET"})
  }
```

lyonlai replied:
give this a try in your test proto and you will see what I'm talking about.

```
option (google.api.http) = {
      get: "/api/query/{a}?test=3"
    };
```

atreya2011 replied:
I see what you mean... then this means when we define the proto, we should add the `?` ourselves? Let me know if I understand your suggestion correctly

```
option (google.api.http) = {
      get: "/api/query/{a}?"
    };
```

atreya2011 replied:
If we have to add the `?` ourselves then I think its a good idea to mention this in the README

lyonlai replied:
nah. I mean. in the generator code, we could tell if there's a query parameter has been specified before and it should be retained. and the query params that we add will need to adjust according to that.

say `get: "/api/query/{a}?test=3"` the query params we add will have `&` appended then with the params.
if just `get: "/api/query/{a}"` then the generator should add a `?` instead

atreya2011 replied:
Aaah! I understand now. So we check this in the Go code and append `?` or `&` as required. Thank you for the detailed explanation 🙇🏼

lyonlai reviewed (changes requested):
it definitely looks nicer on this pass.  there are few changes & cases to cover as per comment. Good job.

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +196,105 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+
+      return { ...acc, ...objectToMerge };
+    },
+    {} as T
```
wouldn't this be `{} as FlattenedRequestPayload`?

lyonlai replied:
well, disregard, T should do the work anyway

lyonlai commented on generator/template.go:
```diff
@@ -229,20 +329,38 @@ func fieldName(r *registry.Registry) func(name string) string {
+			renderURLSearchParamsFn := fmt.Sprintf("${fm.renderURLSearchParams(req, %s)}", fieldsInPath)
+			// prepend "&" if query string is present otherwise prepend "?"
+			if parsedURL.RawQuery != "" {
+				methodURL += "&" + renderURLSearchParamsFn
```
you might want to remove the ending & if there's any before attaching

lyonlai replied:
say if the url in the option is `/paht/to/{a}?test=a&`, then you don't need to add the `&`

atreya2011 replied:
Ah indeed! Thank you for pointing that out 🙇🏼

atreya2011 replied:
I was trying to write a test case for this and `grpc-gateway` throws an error when trying to generate.
`--grpc-gateway_out: unexpected token "&test=1" after segments "api/query/{a}": /api/query/{a}&test=1`
It seems we are not allowed to inject query parameters in the path and therefore I don't think we need the parse the method URL to check for query parameters.
Let me know what you think about this 🙏🏼

lyonlai replied:
this is new to me. looks like an issue on the grpc-gateway. let's leave the logic here coz I don't see there's anything  in the `http.proto` saying it's not permitted?

atreya2011 replied:
Ok sure. I will make the modifications as per your suggestion 👍🏼

lyonlai replied:
hangon. found a problem in  `/api/query/{a}&test=a` shouldn't this be a `?` to start with

lyonlai replied:
should be `/api/query/{a}?test=a`

atreya2011 replied:
Sorry for the typo in my comment, this is the exact error message I got:
`--grpc-gateway_out: unexpected token "?test=a" after segments "api/query/{a}": /api/query/{a}?test=a`

atreya2011 replied:
![Screenshot from 2021-04-27 13-31-41](https://user-images.githubusercontent.com/44151328/116185628-f427a200-a75c-11eb-99fd-0afa27fa9cb4.png)
Leaving a screenshot for clarification 🙇🏼

atreya2011 replied:
Have modified the function as per your suggestion to remove leading `&` if present 🙇🏼

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +196,105 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+type RequestPayload = Record<string, unknown>;
+type FlattenedRequestPayload = Record<string, Primitive | Array<Primitive>>;
+
+function isPlainObject(val: unknown): boolean {
```
can you add a comment here stating the source of the logic?

atreya2011 replied:
Understood. Will do so 🙇🏼

atreya2011 commented:
I have refactored the code according to your suggestions and added a test for checking falsy values as well 🙇🏼

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +196,105 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+
+      if (isPlainObject(value)) {
+        objectToMerge = flattenRequestPayload(value as RequestPayload, newPath);
+      } else if (isPrimitive(value) || isNonEmptyPrimitiveArray) {
```
I think you misunderstood. the 0 & false logic here is more about. 

if say request object `{ a:0, c:false, b: 1)`, then the output query string should be `?b=1`, the current implementation doesn't cover it

atreya2011 replied:
Hmmm... I think we should not filter out zero values from the query string as they may be intentionally added by the user. Let me know your opinion on this.

lyonlai replied:
it's not needed though. and jsonpb also ignores the default value to save space. we should follow the same.

https://developers.google.com/protocol-buffers/docs/proto3#json

atreya2011 replied:
Ah I see.. Thanks for pointing this out... I will modify the flatten function to filter out keys with zero values 🙇🏼

atreya2011 replied:
I have modified the flatten function to filter zero values as well 🙇🏼

atreya2011 commented:
Thank you once again for taking your time in reviewing this. I have furthered fixed the code according to your suggestions 🙇🏼

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +196,129 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+
+      let objectToMerge = {};
+
+      if (isPlainObject(value)) {
+        objectToMerge = flattenRequestPayload(value as RequestPayload, newPath);
+      } else if (
+        isPrimitive(value) &&
+        !isZeroValuePrimitive(value as Primitive)
+      ) {
+        objectToMerge = { [newPath]: value };
+      } else if (isNonEmptyPrimitiveArray) {
+        const nonEmptyPrimitiveArray = value as Primitive[];
+        objectToMerge = {
+          [newPath]: nonEmptyPrimitiveArray.filter(
+            m => !isZeroValuePrimitive(m)
+          )
+        };
+      }
```
I am filtering zero value primitives here

lyonlai replied:
was gonna say, empty values in the array is the case that you shouldn't filter.

atreya2011 replied:
Ah I see hmmm... but if we don't filter them then they will get added to the path as URL query params? Is that OK?

lyonlai replied:
there's no default value in the repeated array. the default value case here only applies for primitive fields. array no need to touch.

e.g. `a: [1,0,1]` will turn  into `a=1&a=0&a=1`,  if you put `a=1&a=1` then in the backend you only have `a:[1,1]`

atreya2011 replied:
Ah I see. I understand what you mean. Will fix this right away 🙇🏼

atreya2011 replied:
> it's not needed though. and jsonpb also ignores the default value to save space. we should follow the same.
> https://developers.google.com/protocol-buffers/docs/proto3#json

I was giving this some consideration. The above documentation also says this
```
 An implementation may provide options to emit fields with default values in the JSON-encoded output.
```
Maybe we can add a flag like golang's grpc gateway plugin `EmitDefault` and if it is set to true, we filter zero value primitives otherwise we don't filter them. Let me know what you think about this 🙏🏼 

Reason is as follows:
Lets take the following payload where `b` & `c` are part of the query parameters `/api/query/a?msg.b=false&c=hello`
```
{
  "a": 10
  "msg": { "b": false },
  "c": hello
}
```

The user may intentionally want to save `false` in the database but if we remove it from query params `/api/query/a?c=hello` and try to read the `msg` object on the server side, it will be of value `nil` and we can't be really sure whether the user wanted to save `false` or `null` to the database.

atreya2011 replied:
This is the option in `grpc-gateway` I am talking about. `EmitUnpopulated`. We can perhaps add a similar option and let the user decide whether they want to filter zero-values or not.

https://grpc-ecosystem.github.io/grpc-gateway/docs/development/grpc-gateway_v2_migration_guide/#we-now-emit-default-values-for-all-fields

lyonlai replied:
grpc-gateway's emitting the default 

https://github.com/grpc-ecosystem/grpc-gateway/blob/master/runtime/marshaler_registry.go#L23

atreya2011 replied:
Yes, as of version 2 they are emitting the default values but as the user of the package we can decide whether we want to enable it to true or false.
Similarly we can have an option here like `EmitUnpopulated` where we can filter default-value fields if set to false and not filter them if set to true.

atreya2011 replied:
I have revised the code as per your suggestions. Maybe `EmitUnpopulated` can be a different PR?

lyonlai replied:
had a second read. what we are talking about here is unmarshalling because the plugin sends a json and grpc gateway uses `protojson` to unmarshal the payload into the generated proto struct. 

Then we were looking at the wrong place. `EmitUnpopulated` affects payload on the way down, not on the way up to grpc-gateway.

So we will always have the default value as per described here.
https://developers.google.com/protocol-buffers/docs/proto3#default

like I mentioned before, it's a thing that's worth a sentence/short paragraph to describe the behaviour.

atreya2011 replied:
Yes, I did a couple of reads and also the above link and came to the same conclusion.
My apologies for the misunderstanding and the long discussion regarding this 🙇🏼 
Should we describe this behaviour in the README or as a comment?

lyonlai replied:
yes please. a section call `Notes` before the `Example` will be good. just number dotted the title and have a few sentences about it.

atreya2011 replied:
Understood. Will do so and update this PR and thank you once again for the patience in reviewing my code and giving me this chance to making this contribution 🙇🏼

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +196,129 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+  return ["string", "number", "boolean"].some(t => typeof value === t);
+}
+
+/**
+ * Checks if given primitive is zero-value
+ * @param  {Primitive} value
+ * @return {boolean}
+ */
+function isZeroValuePrimitive(value: Primitive): boolean {
+  return ["", "0", "false"].some(zv => value.toString() === zv);
+}
```
This function checks if given primitive value is zero-value or not

lyonlai replied:
let's not use toString but do `===` with `||` here.

atreya2011 replied:
Like this ?
```js
value === boolean || value === 0 || value === ""
```

atreya2011 commented on generator/template.go:
```diff
@@ -195,6 +196,129 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+
+/**
+ * Checks if given value is a plain object
+ * source: https://github.com/char0n/ramda-adjunct/blob/master/src/isPlainObj.js
```
Added source link for `isPlainObject`

lyonlai replied:
`This copies the logic from xxx`

atreya2011 replied:
Ah ok! I will modify the comment as per your suggestion 🙇🏼

atreya2011 commented on generator/template.go:
```diff
@@ -229,20 +353,39 @@ func fieldName(r *registry.Registry) func(name string) string {
+			// prepend "&" if query string is present otherwise prepend "?"
+			// trim leading "&" if present before prepending it
+			if parsedURL.RawQuery != "" {
+				methodURL += "&" + strings.TrimRight(renderURLSearchParamsFn, "&")
```
Removing leading `&` if present in URL path here.

lyonlai replied:
shouldn't this be Trimleft?

atreya2011 replied:
ah I am sorry since you mentioned that URL may contain "&" I think it should be 
`methodURL = strings.TrimRight(methodURL, "&") + strings.TrimLeft(renderURLSearchParamsFn, "&")`

lyonlai replied:
it's for the case where methodURL ends with `&`, 

so it should be `strings.TrimRight(methodURL, "&")  + "&" + renderURLSearchParamsFn.

you are missing the `&` in the middle.

atreya2011 replied:
Sorry about the many mistakes and typos. Will fix this 🙇🏼

lyonlai commented on generator/template.go:
```diff
@@ -195,6 +196,123 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
+      if (isPlainObject(value)) {
+        objectToMerge = flattenRequestPayload(value as RequestPayload, newPath);
+      } else if (
+        (isPrimitive(value) && !isZeroValuePrimitive(value as Primitive)) ||
```
let's get `isPrimitive(value) && !isZeroValuePrimitive(value as Primitive)` into a variable `isNonZeroValuePrimitive` to improve readability.

atreya2011 replied:
Sure. I will revise as suggested and my apologies for missing this comment 🙇🏼

atreya2011 replied:
Done 🙇🏼

lyonlai commented on generator/template.go:
```diff
@@ -229,20 +347,39 @@ func fieldName(r *registry.Registry) func(name string) string {
+				fieldsInPath += fmt.Sprintf(`"%s", `, fieldName)
+			}
+		}
+		fieldsInPath = strings.TrimRight(fieldsInPath, ", ") + "]"
```
for constructing strings like an array like this `[a,b,c,d]`.

Would be better to use arrays and join construct the string rather than construct it and trim right. So you can do something like

```golang
fieldsInpath := make([]string, 0, len(matches))

// in the for loop
fieldsInPath := append(fieldsInPath, fieldName)

// out of the loop
fieldsInPath = fmt.Sprintf("[%s]", strings.Join(fieldsInPath, ", "))

```

atreya2011 replied:
Ah! Sorry for my naive implementation. Will revise as per your suggestion 🙇🏼

atreya2011 replied:
Hmm, I tried your example and unfortunately, the fieldsInPath gets generated like this `[a, b, c]` but the expected generated list should be like this `["a", "b", "c"]` 😞 and the tests fail. I am trying to figure out if we can use your suggestion to generate `["a", "b", "c"]` 🙏🏼

atreya2011 replied:
Done 🙇🏼

lyonlai reviewed (changes requested):
Last two comments. This is looking really good. sorry for missing the `fieldsInPath` construction before. 

Also when you finished, can you also please rebase to a single commit and push up?

atreya2011 commented on README.md:
```diff
@@ -42,10 +41,69 @@ Turn Ton logging to stderr. Default to false.
 ### `loglevel`
 Defines the logging levels. Default to info. Valid values are: debug, info, warn, error
 
+### Notes:
```
I have added the notes here. Looking forward to your feedback 🙇🏼

lyonlai replied:
This is too long. The note section is supposed to note the things people unexpected to find, and also supposed to be concise. In the http get case, it is about the field value matches the default value won't be presented in url params. and a short example like the following will be enough . 

`{fieldA: "A", fieldB: "" fieldC:1}` will turned in the `/path?filedA=A&fieldC=1`

atreya2011 replied:
Hmm... ok I will try my best to make it more concise. Thank you for the feedback 🙏🏼

atreya2011 replied:
Sorry for the delay. I have simplified the notes section 🙇🏼

lyonlai commented on README.md:
````diff
@@ -42,10 +41,71 @@ Turn Ton logging to stderr. Default to false.
+### Notes:
+This library generates URL Search Parameters (if required) at runtime for RPCs annotated as GET requests. For example, given the following proto definition
+
+```proto
````
protos in the integration test already covers this

atreya2011 replied:
So maybe we can point out in the notes to look for the integration test example?

lyonlai replied:
I think so. just point a link to the integration test service.proto will do the work.

atreya2011 replied:
Understood. Thanks for the feedback 🙇🏼

atreya2011 replied:
Completed 🙏🏼

lyonlai commented on integration_tests/service.proto:
```diff
@@ -57,6 +57,32 @@ message HttpDeleteRequest {
   int32 a = 1;
 }
 
+message HTTPGetWithURLSearchParamsRequest {
```
It will be good to have the url example here

atreya2011 replied:
Ah ok, you mean the URL example from the notes section in the README?

lyonlai replied:
just a url example will be ok. no need to move the whole thing here.

atreya2011 replied:
I understand, just a simple request and response message and a simple RPC implementation.

atreya2011 replied:
I have modified the tests with the URL example. Please let me know if this good or if I have missed something 🙇🏼

atreya2011 commented:
> Last two comments. This is looking really good. sorry for missing the `fieldsInPath` construction before.
> 
> Also when you finished, can you also please rebase to a single commit and push up?

Also rebased to a single commit 🙇🏼

lyonlai reviewed (approved):
LGTM, thanks for the contribution @atreya2011

atreya2011 commented:
> LGTM, thanks for the contribution @atreya2011
@lyonlai 
Likewise, thank you so much for your patience in reviewing my PR and letting me make this contribution 🙇🏼
